### PR TITLE
Fix: Adjust pie chart title to prevent cropping

### PR DESCRIPTION
**Pull request by nourhenta**
**Summary**
Fixes pie chart title rendering by dynamically reducing font size when the title exceeds available space, ensuring it fits without being cropped.
_Resolves #6242_

**Design Decisions**
Added logic to measure title width using getBBox().

- Introduced a loop to reduce font size (down to a minimum) until the title fits within pieWidth - margin.

- Applied the fix only after title rendering to avoid runtime errors.

- Used .style('font-size', ${fontSize}px) for dynamic font updates.

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://mermaid.js.org/community/contributing.html)
- [x] :computer: have added necessary unit/e2e tests.
- [x] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://mermaid.js.org/community/contributing.html#update-documentation) is used for all new features.
- [x] :butterfly: If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.
n about the content of your PR.

### CODE REVIEW

**Review by sidharthv96 — CHANGES_REQUESTED** [review]
Please add few visual test cases that cover the fixed issue.

**Comment by sidharthv96** [comment]
@nourhenta Why don't we split the title into multiple lines, as the minimum font size is too small currently, and it does lead to cropping after we reach that size.

**Comment by nourhenta** [comment]
Hi @sidharthv96,
I have addressed the requested change and added a few visual test cases that cover the fixed issue as suggested.
Also, I will start working on the improvement you mentioned regarding splitting the title into multiple lines when the font size reaches the minimum, to avoid further cropping.
Let me know if you have any additional feedback. Thanks!

**Comment by nourhenta** [comment]
Hi  @sidharthv96,
I have updated the pie chart title handling as discussed:
-  Long titles now automatically move to a new line if they exceed the maximum width after shrinking to the minimum font size.
How it works now:
- First, the font size is dynamically reduced while the title is too wide.
- If the title is still too wide after reaching the minimum font size, it is automatically split into multiple lines at word boundaries.
- Each line is drawn separately and centered horizontally above the pie chart.
**Visual result:**
image

**Comment by nourhenta** [comment]
Hi @sidharthv96,
The Argos visual differences are expected due to intentional changes in the title rendering:

- 📏 Long titles now automatically shrink font size or split into multiple lines if needed.
- 🛠 Short titles are now explicitly positioned and may have minor font size adjustments compared to before.
- 🎯 These changes were necessary to improve consistency, readability, and accessibility of pie chart titles.

✅ Functionality is working correctly.
✅ Visual changes are minor and aligned with the intended new behavior.

Could you please review and approve the Argos diffs manually?

Thanks!

**Review by sidharthv96 — CHANGES_REQUESTED** [review]
@nourhenta we need to tweak the implementation to utilise existing util functions in mermaid, rather than re-implementing the logic inside pie renderer. 

Can you have a look at `wrapLabel` function here, which already does a lot of what's needed. 

https://github.com/mermaid-js/mermaid/blob/c8daf9aa432139c50fd97b7311b0d0d0741b80d8/packages/mermaid/src/utils.ts#L560-L600

Also the logic needs to change, we should not keep using the lowest font size after the split, but whatever the largest font is possible given the current number of lines. 

image


As you can see in this example, the last line only has a single word, with the whole title rendered in tiny text. It would be better to have a bigger font with more words in the last line.

**Comment by nourhenta** [comment]
Hello  @sidharthv96 ,

-  I updated the pie chart title handling based on your feedback:
- I now use the existing wrapLabel function from utils.ts instead of reimplementing splitting logic inside the pie renderer.
- The font size first tries to shrink to fit in one line, then if still too long, it wraps the text using wrapLabel.
- After wrapping, the largest possible font size is preserved given the number of lines, instead of forcing minimum font size.
- I also adjusted the pie chart positioning and viewBox height dynamically based on the title height to avoid overlap between the title and the pie.

This way the pie stays fully visible, and the title remains readable without being unnecessarily small.

- Also, Argos visual tests are failing because of the expected layout changes (title wrapping, pie repositioning, and margin adjustments).
This failure is normal as the rendering has improved — the chart structure changed slightly to better fit the updated title behavior.
So,could you please review and approve the Argos diffs manually?
Thank you for your guidance! Let me know if there’s anything else you’d like me to adjust. 🙏

File: packages/mermaid/src/diagrams/pie/pieRenderer.ts [hunk 1]
```diff
@@ -8,6 +8,7 @@ import { selectSvgElement } from '../../rendering-util/selectSvgElement.js';
 import { configureSvgSize } from '../../setupGraphViewbox.js';
 import { cleanAndMerge, parseFontSize } from '../../utils.js';
 import type { D3Section, PieDB, Sections } from './pieTypes.js';
+import { wrapLabel } from '../../utils.js';
```
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/pie/pieRenderer.ts:11]: Please combine both these imports.

File: packages/mermaid/src/diagrams/pie/pieRenderer.ts [hunk 2]
```diff
@@ -125,12 +127,81 @@ export const draw: DrawDefinition = (text, id, _version, diagObj) => {
     .style('text-anchor', 'middle')
     .attr('class', 'slice');
 
-  group
+  const titleText = db.getDiagramTitle();
+  const maxAvailableWidth = pieWidth - MARGIN;
+  let fontSize = 25;
+  const minFontSize = 8;
```
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/pie/pieRenderer.ts:133]: 8 looks too small, maybe 10?

```suggestion
  const minFontSize = 10;
```

File: packages/mermaid/src/diagrams/pie/pieRenderer.ts [hunk 3]
```diff
@@ -125,12 +127,81 @@ export const draw: DrawDefinition = (text, id, _version, diagObj) => {
...
+  while (bbox && bbox.width > maxAvailableWidth && fontSize > minFontSize) {
+    fontSize -= 1;
+    tempTitle.remove();
+
+    // Re-wrap at smaller font size
+    wrappedTitle = wrapLabel(titleText, maxAvailableWidth, {
+      fontSize,
+      fontFamily: 'Arial',
+      fontWeight: 400,
+      joinWith: '<br/>',
```
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/pie/pieRenderer.ts:175]: This is repeated from above, please move them into a different variable, so we only need to adjust in one place. 

image

File: packages/mermaid/src/diagrams/pie/pieRenderer.ts [hunk 4]
```diff
@@ -125,12 +127,81 @@ export const draw: DrawDefinition = (text, id, _version, diagObj) => {
...
+      .style('white-space', 'pre-line')
+      .style('font-size', fontSize + 'px');
+
+    lines.forEach((line, idx) => {
+      tempTitle
+        .append('tspan')
+        .attr('x', pieWidth / 2)
+        .attr('dy', idx === 0 ? 0 : '1.2em')
+        .text(line);
+    });
```
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/pie/pieRenderer.ts:195]: This is also duplicated, combine to a function?

File: packages/mermaid/src/diagrams/pie/pieRenderer.ts [hunk 5]
```diff
@@ -125,12 +127,81 @@ export const draw: DrawDefinition = (text, id, _version, diagObj) => {
...
+
+    bbox = (tempTitle.node() as SVGGraphicsElement)?.getBBox(); // <-- JUST this
```
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/pie/pieRenderer.ts:197]: What does the comment signify? 🤔

File: packages/mermaid/src/diagrams/pie/pieRenderer.ts [hunk 6]
```diff
@@ -125,12 +127,81 @@ export const draw: DrawDefinition = (text, id, _version, diagObj) => {
...
+  }
+  const titleHeight = bbox?.height || 0;
+  const TITLE_MARGIN = 10; // add extra margin between title and pie
+  group.attr(
+    'transform',
+    'translate(' + pieWidth / 2 + ',' + (height / 2 + titleHeight / 2 + TITLE_MARGIN) + ')'
```
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/pie/pieRenderer.ts:203]: Please use string template, so it's more readable.

**Review by sidharthv96 — CHANGES_REQUESTED** [review]
We're almost there! 

Seems to be some edge cases to handle?

image
Not enough margin

image
Overlap

We need multiple rendering tests added in `cypress/integration/rendering/pie.spec.ts`, for these various scenarios

**Comment by nourhenta** [comment]
Hello @sidharthv96,
I have addressed all the changes you requested in the previous comments:
1. Merged Imports:
I combined the import statements for cleanAndMerge, parseFontSize, and wrapLabel from utils.js into a single import statement, as requested. This makes the code cleaner and avoids redundancy.
`import { cleanAndMerge, parseFontSize, wrapLabel } from '../../utils.js';`

2. Increased minFontSize:
I updated the minFontSize from 8 to 10, as suggested, to avoid making the title too small. 
`const minFontSize = 10;`

3. Moved Repeated Title Styling Properties to a Variable:

The fontFamily, fontWeight, and joinWith properties used for wrapping the title were repeated in the code. I moved these into a reusable titleWrapConfig object so that we can adjust them in one place.
```
const titleWrapConfig = {
  fontFamily: 'Arial',
  fontWeight: 400,
  joinWith: '<br/>',
};
```
4. Refactored tspan Title Creation Code:
The code to append each line of the title using tspan was duplicated. I created a helper function called createTitle to eliminate this repetition and make the code more readable.
5. Clarified Comment on bbox Calculation:
I removed the comment that was previously on the line with bbox = (tempTitle.node() as SVGGraphicsElement)?.getBBox(); because it seemed unnecessary and redundant. It was meant to mark the point where the title bounding box is calculated, but I agree it wasn’t needed. I’ve cleaned up the code.
6. Improved String Concatenation:
I updated the string concatenation for the transform attribute inside the group.attr() to use a template string for better readability. Here’s the updated code:
```
group.attr(
  'transform',
  `translate(${pieWidth / 2}, ${height / 2 + titleHeight + TITLE_MARGIN})`
);
```
7.finally,The pie chart's position is also adjusted based on the title's height, ensuring that the title never overlaps with the pie, regardless of its length. 
Everything should now be properly refactored and more readable. I also tested the changes and everything appears to be working fine.

- I will work on adding more comprehensive and robust test cases in the pie.spec.ts file to ensure that the pie chart renders correctly across different scenarios.
Let me know if there’s anything else you'd like me to adjust. 🙏

**Comment by nourhenta** [comment]
Hello  @sidharthv96,
I have added the requested rendering tests in `cypress/integration/rendering/pie.spec.ts`
I ran all tests locally and confirmed they pass .
Let me know if you need any adjustments. Thank you!

File: cypress/integration/rendering/pie.spec.ts [hunk 7]
```diff
@@ -82,4 +82,30 @@ describe('pie chart', () => {
...
+  });
+
+  it('wraps very long title into multiple lines', () => {
+    imgSnapshotTest(`
+      pie
+        title Short
+        "A" : 60
+        "B" : 40
+    `);
+  });
```
Comment (sidharthv96) [inline cypress/integration/rendering/pie.spec.ts:110]: All these are short titles?

**Review by sidharthv96 — CHANGES_REQUESTED** [review]
Code looks good!
We need to update the tests. 

Also add check to verify that a veeeeery long string would be split with `-` as well. 

image

File: cypress/integration/rendering/pie.spec.ts (hunk 7 above)
Comment (nourhenta) [reply cypress/integration/rendering/pie.spec.ts:110]: Thank you for pointing that out — you're absolutely right!
The same short title was mistakenly used in all three tests.
I’ll correct them now by providing:
- A short title for the first test (as intended),
- A medium-length title to check font shrinking,
- A very long title to check line wrapping.

**Comment by sidharthv96** [comment]
image

This issue is still present, an extra padding on top even when there is no title.

**Comment by sidharthv96** [comment]
The shift is still present.

image

These needs to be equal

image

---

Please don't force push, as it makes it harder for me to see what was changed to resolve the issue.
